piccolo/query/mixins.py

```diff
@@ -613,12 +616,132 @@ def __str__(self):
         self._group_by = GroupBy(columns=columns)
+
+
+class OnConflictAction(str, Enum):
+    do_nothing = "DO NOTHING"
+    do_update = "DO UPDATE"
+
+
+@dataclass
+class OnConflict:
+    target: t.Optional[t.Sequence[t.Union[str, Column]]] = None
+    action: t.Optional[OnConflictAction] = None
+    values: t.Optional[
+        t.List[t.Union[Column, t.Tuple[t.Union[str, Column], t.Any]]]
+    ] = None
+
+    @property
+    def target_string(self) -> str:
+        assert self.target
+
+        def to_string(value) -> str:
+            if isinstance(value, Column):
+                return f'"{value._meta.name}"'
+            elif isinstance(value, str):
+                return value
+            else:
+                raise ValueError("OnConflict.target isn't a valid type")
+
+        columns_str = ", ".join([to_string(i) for i in self.target])
+        return f"({columns_str})"
+
+    @property
+    def action_string(self) -> QueryString:
+        action = self.action
+        if isinstance(action, OnConflictAction):
+            if action == OnConflictAction.do_nothing:
+                return QueryString(OnConflictAction.do_nothing.value)
+            elif action == OnConflictAction.do_update:
+                values = []
+                query = f"{OnConflictAction.do_update.value} SET"
+
+                if not self.values:
+                    raise ValueError("No values specified for `on conflict`")
+
+                for value in self.values:
+                    if isinstance(value, Column):
+                        column_name = value._meta.db_column_name
+                        query += f' "{column_name}"=EXCLUDED."{column_name}"'
+                    elif isinstance(value, tuple):
+                        column = value[0]
+                        value_ = value[1]
+                        if isinstance(column, Column):
+                            column_name = column._meta.db_column_name
+                        elif isinstance(column, str):
+                            column_name = column
+                        else:
+                            raise ValueError("Unsupported column type")
+
+                        query += f' "{column_name}"={{}}'
+                        values.append(value_)
+
+                return QueryString(query, *values)
+
+        raise ValueError("OnConflict.action isn't a valid type")
+
+    @property
+    def querystring(self) -> QueryString:
+        query = " ON CONFLICT"
+        values = []
+
+        if self.target:
+            query += f" {self.target_string}"
+
+        if self.action:
+            query += " {}"
+            values.append(self.action_string)
+
+        return QueryString(query, *values)
+
+    def __str__(self) -> str:
+        return self.querystring.__str__()
+
+
+@dataclass
+class OnConflictDelegate:
+    """
+    Used with insert queries to specify what to do when a query fails due to
+    a constraint::
+
+        .on_conflict(action='DO NOTHING')
```

Comment:
I can't try PR right now but it looks great to me. This should also work for duplicate M2M entries like this
```python
if unsaved:
  await rows[0].__class__.insert(*unsaved).on_conflict(action='DO NOTHING').run()
```

Reply:
Yes, hopefully 🤞

Reply:
I just checked this and it works.

Reply:
Thanks 👍